### ISSUE TITLE
Sort order of exported entries. By citekey or zotero order as on screen.

### ISSUE THREAD

**Issue by JohnLukeBentley** [retorquere/zotero-better-bibtex#766]
A feature request to provide an option to sort the exported entries either by citekey or the order as found in the current view of zotero.

**Comment by retorquere** [comment]
Turn out not only can I not affect the order in which Zotero hands translators the references to be exported, I don't *want* this to be possible because it would mess up my test framework something bad.

The references are offered to translators either in order of creation or last-modified date (I forget which); you can sort by those in Zotero by adding those columns, and that would mean you'd get the exported references in the same order as they're present in the export.

**Comment by retorquere** [comment]
There is no difference between "initial order" and "hand-off order" -- the translators get handed the references selected for export in "hand-of order", which is probably only sorted by happenstance. I don't think Zotero spends any effort sorting them, just that the DB query hands them in a certain order, which is probably last-modified on the itemID. Just guessing here.

BBT exports references in the order they're handed to the translators. In my tests, I test the actual import to a known-good match. If the order in which references are handed to translators would vary, my match files wouldn't just be a simple string-match for the output.

It would mess up my tests, but not in a way that I could not fix. But it may very well also mess up Zotero's tests.

**Comment by JohnLukeBentley** [comment]
On the current order Zotero uses ... it looks like (from experiment) "Date Added" is the sort order (that's what I get when I export to BBT).

> Just guessing here.

It looks like you've been here before and even had your hand in the Zotero guts to do some sorting: https://forums.zotero.org/discussion/60375/sort-items-zotero-offers-to-translator; and https://github.com/zotero/zotero/pull/1051/files#r68575230 where it looks like you (then) sorted on "id" (an internal id number?).  :)

Perhaps I don't understanding the meaning of "last-modified on the itemID". 

> There is no difference between "initial order" and "hand-off order" 

Indeed by "initial order" I meant "the initial order BBT receives the entries, being, in other words, the order Zotero hands it off to BBT". That is, for future comparison by BBT which it now looks like is against a "known-good match".

> If the order in which references are handed to translators would vary, my match files wouldn't just be a simple string-match for the output.

So if I understand this correctly: what's important to you (as Translator tester) is an order on a particular field because the known-good tests where previously created on that particular field (and we've discovered, I think, that the particular field is "Date Added").

If that's correct then if Zotero could export by any other field (e.g. current UI order) so long as you, as translator tester, could continue to sort by the field used when you created your known-good tests ... which would seem available to you so long Zotero continued to make available sort by Date Added in the UI.


Anyway where do I shove that ruby text in https://github.com/retorquere/zotero-better-bibtex/issues/590#issuecomment-341888141 ? That is, is that:

* To go somewhere in the Zotero environment;
* ... the BBT environment;
* Or an example of post processing sorting of whatever *.bib file I've dumped entries into (using a ruby install of my own)

?

**Comment by retorquere** [comment]
It is correct that, as long as Zotero reliably hands off the references in a certain order, my test suite would continue to work. I'd have to go re-sort all my match files, which would not be fun, but it'd be possible. Sorting them myself would undesirable because I'm handed the references and am expected to push out the results in a streaming fashion. To sort I'd have to buffer both input and output, which would not only make BBT slower, it'd use more memory (and it's already using a lot), and I'd be fighting Zotero's architecture. Not things I'm keen on doing. 

Sorting by the visual representation for the purposes of my tests would be non-trivial for me since I can't click on stuff during my tests. I drive Zotero by its internal APIs. Changing the tests so they'd interact with the UI is *not* something I'd want to do.

The script I posted would have to be ran from the command line. What platform are you on?

**Comment by retorquere** [comment]
(to be clear, I don't oppose the sorting - I like the idea of it, aesthetically, but I any time I spend on BBT cannibalizes the time I really ought to be spending on work, prepping for my PhD, and time I really should be spending with my family. BBT is already taking a hugely disproportionate chunk of that balance)

**Comment by JohnLukeBentley** [comment]
I previously wrote a long response ... but somehow it was lost into the aether... I begin again ...

I want to see if there's a path that minimizes your time on this issue ...

In principle there'd be three ways to achieve a user desired sort order:

1. Zotero sorts it before handing it off to BBT;
2. BBT sorts it;
3. The (technically sophisticated) user sorts it.

Commenting on those in reverse order ....

With thanks you've already provided a type 3 solution. The ruby script in #590. Once I understood what this was, a type 3 solution, I've embarked on my own solution. I'm part way through writing a small java console app to sort a *.bib file.

You've explained why a type 2 solution, BBT sorting it, is undesirable. It's too expensive in terms of memory.

You've mentioned, with respect to a type 1 solution, Zotero sorting it: that your tests need (for time saving purposes) to maintain their current order; that you don't want to click on a UI before having to run your tests.

However, it seems to me a type 1 solution, Zotero sorting it, is in-principle achievable without impacting your testing workflow.

Zotero could sort according to the UI and, when you download (or compile), the latest version of Zotero it can be made to come, by default, with the current sort order. That way you can run your tests as they are. If you spent some time click on the UI, as a user of Zotero, you'd need only click on the columns to restore your desired test order (Date Added or whatever) before you run your tests.

In addition, if you'd rather not even have to think about sort order before you run your tests, the Zotero API could provide a sort order that overrides the UI set sort order.

Do you see any problems with something like that? If not I'd make a feature request on the Zotero forums (for some dev apart from yourself to-do).

In the meantime, and regardless of the above, I'll finish my own Java sort tool and, then, provide you with the list of EDTF dates (in #590).

**Comment by retorquere** [comment]
> In principle there'd be three ways to achieve a user desired sort order:

Yes.

> I'm part way through writing a small java console app to sort a *.bib file.

To each their own 😆.  I remember doing Java but they're not wholly fond memories. I spend most of my time in scripting languages these days -- ruby, haskell, python in a pinch. Oh and javascript of course. Not my favorite of the bunch (there's a good reason for books like "JavaScript: The Good Parts") but it's usable.

> You've explained why a type 2 solution, BBT sorting it, is undesirable. It's too expensive in terms of memory.

Plus it'd mean a major restructuring of the BBT exporters. The memory overhead is the downside to you, the user. The restructuring is the downside to me. I've only just gotten close to getting BBT5 to work. Another major restructuring is not high on my bucket list.

> You've mentioned, with respect to a type 1 solution, Zotero sorting it: that your tests need (for time saving purposes) to maintain their current order; that you don't want to click on a UI before having to run your tests.

It's not (just) that I don't want to, I can't. The tools I use were meant to drive a browser, not a XUL interface. The tool (Selenium) can't interact with the UI of Zotero. I'm loading a second plugin to add a test API to Zotero just so I can drive its internal behavior. Zotero will react to these (it's pretty fun to see actually when the test suite is running) but I can't actually click anything, just call the stuff that would have happened *if* the user had clicked something. If that makes any sense.

> However, it seems to me a type 1 solution, Zotero sorting it, is in-principle achievable without impacting your testing workflow.

It would impact my current workflow, but I can probably work around it. I'd write something so my test suite would sort the returned export just before comparing it. This would probably work. I don't know how much work that'd be, exactly, but nothing on the order of option 2 of your list.

> Zotero could sort according to the UI and, when you download (or compile), the latest version of Zotero 

Download. I always test against the latest released version without (the possibility of) changes.

> it can be made to come, by default, with the current sort order. That way you can run your tests as they are.

That... could work yeah I think?

> If you spent some time click on the UI, as a user of Zotero, you'd need only click on the columns to restore your desired test order (Date Added or whatever) before you run your tests.

The Zotero profile I test against gets re-created each test run to ensure reproducability. It's not my production profile. No problems there.

> In addition, if you'd rather not even have to think about sort order before you run your tests, the Zotero API could provide a sort order that overrides the UI set sort order.

> Do you see any problems with something like that? If not I'd make a feature request on the Zotero forums (for some dev apart from yourself to-do).

No problems with that.

But do you want the references after export to be sorted a specific way, or do you just want them to be in display order? Because if its the latter, you can just sort by date-added in the Zotero UI, and you'd be done.

> In the meantime, and regardless of the above, I'll finish my own Java sort tool and, then, provide you with the list of EDTF dates (in #590).

Cool. I expect to finish AUX scanning (#758) today, and after that, plus this (#590), I will release BBT5 as a proper release.

**Comment by peterahrens** [comment]
If it helps anyone who comes to this issue looking for a way to sort their bibtex:
`bibtool` is a powerful tool for formatting your bibtex. You can sort in any order you want (and align and reformat and suppress/reorder/reformat/filter fields and more!)

bibtool is a CTAN package: https://ctan.org/pkg/bibtool

Here is the manual (sorting on page 66): http://mirrors.concertpass.com/tex-archive/biblio/bibtex/utils/bibtool/bibtool.pdf

TL;DR: `bibtool -s key bibliography.bib`
